core/src/main/java/org/pentaho/di/connections/vfs/provider/ConnectionFileNameParser.java

```diff
@@ -22,36 +22,203 @@
+ *
+ * <p>
+ * Parsing a URI validates its syntax and transforms it into canonical form:
+ * <ul>
+ *   <li>validates the scheme is "pvfs" and is followed by "://"</li>
+ *   <li>"\" file path separators in the path sections are transformed into the canonical file path separator, "/"</li>
+ *   <li>empty path segments, e.g. "//", are removed</li>
+ *   <li>"." path segments are removed</li>
+ *   <li>".." path segments are validated and resolved</li>
+ *   <li>a path with a trailing slash is recognized as a folder</li>
+ *   <li>any percent-encoded characters in path segments which are not reserved are decoded; currently, only the "%"
+ *       character is reserved</li>
+ * </ul>
+ *
+ * <p>
+ * See {@link #SPECIAL_CHARACTERS_FULL_SET} for examples of special characters allowed (without encoding) in connection
+ * names and in path segments. Spaces and international characters are also supported.
+ *
+ * @see VFSConnectionDetails#getRootPath()
+ * @see org.pentaho.di.connections.vfs.VFSConnectionManagerHelper#getResolvedRootPath(VFSConnectionDetails)
+ */
 public class ConnectionFileNameParser extends AbstractFileNameParser {
-  @Override public FileName parseUri( VfsComponentContext vfsComponentContext, FileName fileName, String uri )
+  private static final String[] SCHEMES = new String[] { ConnectionFileProvider.SCHEME };
+
+  private static final char[] RESERVED_CHARS = new char[] { '%' };
+
+  /**
+   * Full set of special characters the connection name can be.
+   * Only excluding the character '/', '\' and '%', which are the file path delimiter, the alternative file path
+   * delimiter, and the URL-encoding escape character, respectively.
+   * <p/>
+   * Special characters in this context are characters that generally have to be encoded
+   * otherwise the use of {@link java.net.URI} will throw {@link java.net.URISyntaxException} when during
+   * object instantiation or subsequent method calls.
+   */
+  public static final String SPECIAL_CHARACTERS_FULL_SET = "!\"#$&'()*+,-.:;<=>?@[] \t\n\r^_`{|}~";
+
+  private static ConnectionFileNameParser instance;
+
+  @NonNull
+  private final ConnectionFileNameUtils connectionFileNameUtils;
+
+  public ConnectionFileNameParser() {
+    this( ConnectionFileNameUtils.getInstance() );
+  }
+
+  public ConnectionFileNameParser( @NonNull ConnectionFileNameUtils connectionFileNameUtils ) {
+    this.connectionFileNameUtils = Objects.requireNonNull( connectionFileNameUtils );
+  }
+
+  @NonNull
+  public static ConnectionFileNameParser getInstance() {
+    if ( instance == null ) {
+      instance = new ConnectionFileNameParser();
+    }
+
+    return instance;
+  }
+
+  /**
+   * Gets an array of reserved characters.
+   * <p>
+   * This list is in sync with the results of {@link #encodeCharacter(char)}.
+   * This is needed in this format for use by {@link UriParser#encode(String, char[])}.
+   * <p>
+   * Currently, only the "%" character is reserved.
```

Comment:
for maintainability and not constantly update with every change, you can just use `{@value  #RESERVED_CHARS} `or` {@link #RESERVED_CHARS}`